src/main/java/com/commercetools/sync/services/impl/CategoryServiceImpl.java

```diff
@@ -36,38 +34,25 @@
 public final class CategoryServiceImpl extends BaseServiceWithKey<CategoryDraft, Category, CategorySyncOptions,
     CategoryQuery, CategoryQueryModel, CategoryExpansionModel<Category>> implements CategoryService {
 
-    private static final String CATEGORY_KEY_NOT_SET = "Category with id: '%s' has no key set. Keys are required for "
-        + "category matching.";
-
     public CategoryServiceImpl(@Nonnull final CategorySyncOptions syncOptions) {
         super(syncOptions);
     }
 
     @Nonnull
     @Override
-    public CompletionStage<Map<String, String>> cacheKeysToIds() {
-
-        if (isCached) {
-            return CompletableFuture.completedFuture(keyToIdCache);
-        }
-
-        final Consumer<List<Category>> categoryPageConsumer = categoriesPage ->
-            categoriesPage.forEach(category -> {
-                final String key = category.getKey();
-                final String id = category.getId();
-                if (isNotBlank(key)) {
-                    keyToIdCache.put(key, id);
-                } else {
-                    syncOptions.applyWarningCallback(new SyncException(format(CATEGORY_KEY_NOT_SET, id)),
-                        category, null);
-                }
-            });
-
-        return CtpQueryUtils
-            .queryAll(syncOptions.getCtpClient(), CategoryQuery.of(),
-                categoryPageConsumer)
-            .thenAccept(result -> isCached = true)
-            .thenApply(result -> keyToIdCache);
+    public CompletionStage<Map<String, String>> cacheKeysToIds(@Nonnull final Set<String> categoryKeys) {
+        return cacheKeysToIds(
+                categoryKeys, keysNotCached -> CategoryQuery
+                    .of()
+                    .withPredicates(buildCategoryKeysQueryPredicate(keysNotCached)));
+    }
+
+    QueryPredicate<Category> buildCategoryKeysQueryPredicate(@Nonnull final Set<String> categoryKeys) {
```

Comment:
would it be better to make it private ?

Reply:
ah, good hint. wanted to write a test for it but forgot :) 
actually there are no UnitTests for this class, should i add some?

Reply:
Not for now, probably it's covered by integration tests.